### PR TITLE
Diagnostic: extend from_env function

### PR DESCRIPTION
**Pull request by BelovDV**
Implements major part of #51.

This pr adds error type for `from_env()` and returns possibility to check if fds refer to pipe as optional feature.

### CODE REVIEW

File: src/unix.rs
Comment (weihanglo) [inline src/unix.rs:100]: Would it be more intuitive with `Result<Option<Client>, ErrFromEnv>`? So that we can return `Err(…)` instead of `Some(Err(…))`.

Or even more, just `Result<Client, ErrFromEnv>` and return `Err(ParseEnvVar)` for the old `None` code path.

File: src/unix.rs
Comment (weihanglo) [inline src/unix.rs:92]: This is a bit unfortunate. Previously if people is using fifo-style jobserver, they don't need to read from pipe. Now they need both.

File: src/unix.rs
Comment (weihanglo) [inline src/unix.rs:120]: ditto

File: src/unix.rs
Comment (weihanglo) [inline src/unix.rs:405]: Could you provide a lint to the reference documentation or code? Thank you!

**Review by weihanglo — COMMENTED** [review]
This looks good to me in general. Let's see other's thought on the change :)

File: src/unix.rs
Comment (BelovDV) [reply src/unix.rs:100]: Current `None` isn't actual error. Wouldn't it be less intuitive to use `Err` in this case? I'm not sure.

As for `Result<Option<Client>, ErrFromEnv>` - it's better, than my variant, if `Option` will remain here - I'll change to this.

File: src/unix.rs
Comment (BelovDV) [reply src/unix.rs:92]: I'll check, if both `fifo:PATH` and `R,W` can be provided, but even if they can and will. Both clients will be created, but only one `from_fifo` will be used. Is smth bad in calling `from_pipe` after succeeded `from_fifo`?

File: src/unix.rs
Comment (weihanglo) [reply src/unix.rs:100]: I proposed removing `Option` because [there](https://github.com/alexcrichton/jobserver-rs/pull/52/files#diff-52b718f0ee2e4e8580f7155fe871a4267a6427607bf486005bed17325f066a1dR95) with `(None, None)` we actually convert into `ParseEnvVar`. I feel like they are a bit inconsistent. Why is it an error here but not here?

File: src/unix.rs
Comment (BelovDV) [reply src/unix.rs:405]: This is partially reverted #6, copy of code from old version. Or you mean to insert this in code?

File: src/unix.rs
Comment (weihanglo) [reply src/unix.rs:92]: >  Is smth bad in calling from_pipe after succeeded from_fifo?

Not really. Just an idea that if we don't need it we don't create it. It also reduces syscalls. Anyway, I am not the maintainer of this project. Just someone cares about it :)

File: src/unix.rs
Comment (weihanglo) [reply src/unix.rs:405]: I see. Thanks for the reference!

**Comment by BelovDV** [comment]
Thanks for review, I'll wait other's comments.

File: src/unix.rs
Comment (weihanglo) [reply src/unix.rs:100]: Sounds reasonable! I am sold 😆

Just realized there is no button "Mark as Resolved" for this thread even it was initiated by me 🤔

File: src/lib.rs
Comment (NobodyXu) [inline src/lib.rs:162]: Nit: Having os specific exposed doesn't sound great
I wonder if it is a better idea to use `io::Error` here since the return of error here is to improve logging/err msg on failure.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (BelovDV) [reply src/lib.rs:162]: Some errors aren't `io::Error`, wouldn't such error type mislead? I've chose from `imp::ErrFromEnv` and `{ description: &'static str, additional_info: String }`.

The reason of current choice. Os specific errors can be either printed or passed to `match` to be processed. Default `Debug` can be changed. If user want to process `PlatformSpecific`, probably it's better to explicitly state by using `cfg`, that code will deal with platform specific behavior.

If it would be enough in most cases to separate `IsNotConfigured` from other errors, I'd be sure in current variant, but unix `InvalidDescriptor`, according to gnu make, should be considered as a warning, not an error. So if os specific shouldn't be in public interface - I'll change this.

Comment (NobodyXu) [reply src/lib.rs:162]: > Some errors aren't io::Error, wouldn't such error type mislead? I've chose from imp::ErrFromEnv and { description: &'static str, additional_info: String }.

`io::ErrorKind::Others` can be used for these kind of errors and other crates I've seen also use `io::Error` for internal errors, so not exactly misleading.

For example, `tokio::task::JoinError` implements `impl From<JoinError> for std::io::Error`.

[`std::thread::Builder::spawn`](https://doc.rust-lang.org/std/thread/struct.Builder.html#method.spawn) also uses `io::Error` even if creating a new thread isn't an I/O.

> The reason of current choice. Os specific errors can be either printed or passed to match to be processed. Default Debug can be changed. If user want to process PlatformSpecific, probably it's better to explicitly state by using cfg, that code will deal with platform specific behavior.

Does the user really need that much flexibility?
I think if you want to create a new error type for `jobserver`, you'd better create a new `struct Error` that does not expose any public fields for the sake of backwards compatibility and instead only expose methods to return `ErrorKind` and implement `std::error::Error::source` for user to access the underlying source if possible.

> If it would be enough in most cases to separate IsNotConfigured from other errors, I'd be sure in current variant, but unix InvalidDescriptor, according to gnu make, should be considered as a warning, not an error. So if os specific shouldn't be in public interface - I'll change this.

IMO `io::ErrorKind` already provides enough kinds to classify the requirements in #51 

> Error or warn if the jobserver descriptors are open/accessible, but corrupted, i.e. reopened by someone else and no longer refer to pipes 

IMO [`io::ErrorKind::InvalidData`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.InvalidInput) can be used for this.

File: src/lib.rs
Comment (NobodyXu) [inline src/lib.rs:169]:
```suggestion
impl std::fmt::Display for ErrFromEnv {
    // ...
}

impl std::error::Error for ErrFromEnv {
    fn source(&self) -> Option<&dyn std::error::Error> {
        match self {
            ErrorFromEnv::PlatformSpecific { err } => Some(err),
            _ => None,
        }
    } 
}
```

Or you can use [`thiserror`](https://docs.rs/thiserror) for convenience, but since jobserver is used in cargo they might not want this.

File: src/unix.rs
Comment (NobodyXu) [inline src/unix.rs:31]: The filename should be `PathBuf` just in case of non-utf8 characters (jobserver currently uses `env::var` so it is guaranteed to be valid utf8, but IMO it should use `env::var_os` instead for non utf-8 filename).

```suggestion
    OpenFile { filename: PathBuf, err: std::io::Error },
```

File: src/unix.rs
Comment (NobodyXu) [inline src/unix.rs:28]: You should also implement `std::fmt::Display` and `std::error::Error`.

File: src/unix.rs
Comment (NobodyXu) [inline src/unix.rs:401]:
```suggestion
    #[cfg(feature = "check_pipe")]
```

IMHO this check should not be optional

File: src/windows.rs
Comment (NobodyXu) [inline src/windows.rs:144]: Wouldn't it be better to return `io::Error` here, which will give much better error message?

File: src/unix.rs
Comment (NobodyXu) [inline src/unix.rs:32]: It's really confusing what each field is for and what is the reason for the failure.

If you don't want to use `io::Error`, you should at least include a `reason` on why the fd is rejected: Is it non-existent or is it not a pipe?

```suggestion
    InvalidDescriptor { read: std::os::fd::RawFd, write: std::os::fd::RawFd, reason: &'static str },
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (BelovDV) [reply src/lib.rs:162]: I didn't know enough about `io::Error`. Thanks, I'll change to this.

File: src/unix.rs
Comment (BelovDV) [reply src/unix.rs:401]: It was removed with one reason. And there isn't such check in gnu make.

I'm not sure about default behavior though.

Comment (NobodyXu) [reply src/unix.rs:401]: [POSIX jobserver](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html) requires the jobserver on Unix to be implemented with either named pipe or fifo, so if they pass something else, it's incorrect and should be reject, at least that's how I interpret the spec.

File: Cargo.toml
Comment (NobodyXu) [inline Cargo.toml:18]: This is also not a good idea since features should be additive.

With this do_not_check_pipe feature,if one of the crate enables this, then  every other crate is opt out of checking.

File: src/unix.rs
Comment (NobodyXu) [inline src/unix.rs:103]: Use ok_or_else to only create io::Error on failure, since io::Error::new does heap allocation internally.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (BelovDV) [reply Cargo.toml:18]: Otherwise if one enables check then one uses smth different from pipe will fail.

**Comment by alexcrichton** [comment]
Given how low-level this crate is I don't think that the Cargo feature added here will be all that useful since if it's in use it's buried so deep and will be difficult to configure.

Can you explain a bit more what the error handling around pipe detection is doing? Why is `fcntl` not sufficient and what is `fstat` catching otherwise? Is that catching real issues happening in the wild and allowing callers to recover?

**Comment by petrochenkov** [comment]
>I don't think that the Cargo feature added here will be all that useful

+1

> Can you explain a bit more what the error handling around pipe detection is doing? Why is `fcntl` not sufficient and what is `fstat` catching otherwise?

The pipe check is the same thing that jobserver-rs did before https://github.com/alexcrichton/jobserver-rs/pull/6.

> Is that catching real issues happening in the wild and allowing callers to recover?

It catches a real issue that happens in the wild.
The caller won't be able to recover, but it will be able to report a useful error that will allow the "second order caller" (build engineer) to fix the build environment.

File: Cargo.toml
Comment (petrochenkov) [reply Cargo.toml:18]: If checking or not checking the pipe-ness should be configurable, then it should probably be a runtime option passed to `from_env`.

Not checking is the GNU Make behavior, checking catches errors in practice.

File: src/lib.rs
Comment (petrochenkov) [inline src/lib.rs:272]: I'd suggest to keep the simple version of the function alone, and implement it in terms of the "extended" function.
It will also avoid breaking changes to the crate.
Not everyone needs the new diagnostics immediately.

**Comment by alexcrichton** [comment]
If this is reverting back, is the use case in https://github.com/alexcrichton/jobserver-rs/pull/6 no longer valid?

For "catching things in the wild" I meant moreso the new logic to detect a valid file descriptor, less so "should this return errors at all" because that's naturally a "sure this should happen when the effort is put in".

File: src/lib.rs
Comment (petrochenkov) [inline src/lib.rs:162]: @weihanglo 
Could you try this interface in cargo before it's in a "stable" release of jobserver-rs?

It's possible that in practice we'll need to discern between different kinds of erros (e.g. "descriptors are closed" and everything else) to avoid bothering users with warnings too often.

This PR, however, merges all the errors into a single `io::Error` with multiple vague  `ErrorKind`s.
I'm not sure users will be able to use it reliably without relying on "string typing" (inspecting string descriptions returned by errors) or implementation details ("we know that closed descriptors are `ErrorKind::InvalidData` by reading source code of jobserver-rs").
(I don't like this part and I wish I read this PR in detail before its merge.)

Comment (weihanglo) [reply src/lib.rs:162]: Sure. I'll try taking a look this week.

Comment (NobodyXu) [reply src/lib.rs:162]: > implementation details ("we know that closed descriptors are ErrorKind::InvalidData by reading source code of jobserver-rs").

Maybe jobserver can have the error kind documented on docs.rs?

Comment (weihanglo) [reply src/lib.rs:162]: Sorry for the late response. I think the change works fine with Cargo. Cargo can just [ignore the jobserver if there are bad pipes](https://github.com/rust-lang/cargo/blob/7bf43f028ba5eb1f4d70d271c2546c38512c9875/src/cargo/util/config/mod.rs#L258-L262). No harm to get this patch released on Cargo side I believe :)

**Pull request merged**
